### PR TITLE
Implement CheckBytes on ArchivedOptionBox and ArchivedOptionNonZeroN, plus `ArchivedOptionNonZeroN::take()`

### LINKED ISSUE #368: add CheckBytes impls for archived niche types
`ArchivedOptionNonZero*` and `ArchivedOptionBox` are currently lacking a `CheckBytes` implementation.

### LINKED ISSUE #474: Add API to set niched `Option<NonZero*>` types to `None`
Current [ArchivedOptionNonZeroU32](https://docs.rs/rkyv/latest/rkyv/niche/option_nonzero/struct.ArchivedOptionNonZeroU32.html#) supports limited mutation using its multiple `mut` or `insert` methods. But there is no way for us to reset it None.

Similar problem exists with `ArchivedOption` but it being an enum, we can easily set it to `ArchivedOption::None`. But no such option is available in case of `ArchivedOptionNonZeroU32`.

Please provide something like a `take` method and set with `Option` method so that we can set the value to `None` if needed.

### PR DESCRIPTION
**Pull request by novacrazy**
As the title says.

Closes #474
Closes #368

The `ArchivedOptionNonZeroN::take()` is a bit extra, but I was in the file.

### CODE REVIEW

**Comment by novacrazy** [comment]
I need to test the ArchivedOptionBox more.

**Comment by novacrazy** [comment]
Is it okay to leave that test where it is or should I move it to `rkyv_test`? 

I did have to reimplement it to bypass `ArchivedBox::verify`, as that fails with null pointers required for the `None` variant. It's a bit awkward, but should work fine.

File: rkyv/src/niche/option_box.rs [hunk 1]
```diff
@@ -20,6 +20,54 @@ pub struct ArchivedOptionBox<T: ArchivePointee + ?Sized> {
...
+    where
+        T: ArchivePointee + CheckBytes<C> + LayoutRaw + ?Sized,
+        T::ArchivedMetadata: CheckBytes<C>,
+        C: Fallible + ArchiveContext + ?Sized,
+        C::Error: Error,
+    {
+        #[inline]
+        fn verify(&self, context: &mut C) -> Result<(), C::Error> {
+            if self.inner.is_null() {
+                return Ok(());
```
Comment (djkoloski) [inline rkyv/src/niche/option_box.rs:63]: Nit: I would prefer `if .. { Ok(()) } else { self.inner.verify(context) }` to avoid the explicit `return`.

**Review by djkoloski — APPROVED** [review]

**Comment by djkoloski** [comment]
> Is it okay to leave that test where it is or should I move it to `rkyv_test`?

In the future I'd like to move the tests out of rkyv_test and back into the main codebase, so this is actually a step in the right direction.

**Comment by novacrazy** [comment]
Should be good to go. You're welcome to squash this.

**Comment by djkoloski** [comment]
Thanks for the PR! Sorry for the delay!

**Pull request merged**